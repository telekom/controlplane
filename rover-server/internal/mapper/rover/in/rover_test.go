// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package in

import (
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	roverv1 "github.com/telekom/controlplane/rover/api/v1"
)

var _ = Describe("Rover Mapper", func() {
	Context("MapRover", func() {
		It("must map the api format to the internal format", func() {
			input := apiRover
			output := &roverv1.Rover{}

			err := MapRover(input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must return an error if the input rover is nil", func() {
			output := &roverv1.Rover{}

			err := MapRover(nil, output)

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("input rover is nil"))
		})

	})

	Context("MapExposures", func() {
		It("must map exposures correctly", func() {
			input := apiRover
			output := &roverv1.Rover{}

			err := mapExposures(input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("MapSubscriptions", func() {
		It("must map subscriptions correctly", func() {
			input := apiRover
			out := &roverv1.Rover{}

			err := mapSubscriptions(input, out)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), out)
		})
	})

	Context("MapRequest", func() {
		It("must map a RoverUpdateRequest to a Rover correctly", func() {
			output, err := MapRequest(roverUpdateRequest, resourceIdInfo)

			Expect(err).To(BeNil())

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must return an error if the input rover is nil", func() {
			output, err := MapRequest(nil, resourceIdInfo)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("input rover is nil"))
		})

	})
})
