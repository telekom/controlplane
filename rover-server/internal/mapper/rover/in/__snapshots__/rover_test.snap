
[Rover Mapper MapRover must map the api format to the internal format - 1]
&v1.Rover{
    TypeMeta:   v1.TypeMeta{},
    ObjectMeta: v1.ObjectMeta{},
    Spec:       v1.RoverSpec{
        Zone:           "zone",
        IpRestrictions: (*v1.IpRestrictions)(nil),
        ClientSecret:   "",
        Exposures:      {
            {
                Api: &v1.ApiExposure{
                    BasePath:  "/eni/distr/v1",
                    Upstreams: {
                        {URL:"https://httpbin.org/anything", Weight:0},
                    },
                    Visibility: "World",
                    Approval:   v1.Approval{
                        Strategy:     "Simple",
                        TrustedTeams: nil,
                    },
                    Transformation: (*v1.Transformation)(nil),
                    Traffic:        v1.Traffic{},
                    Security:       (*v1.Security)(nil),
                },
                Event: (*v1.EventExposure)(nil),
            },
        },
        Subscriptions: {
            {
                Api: &v1.ApiSubscription{
                    BasePath:       "/eni/distr/v1",
                    Organization:   "",
                    Transformation: (*v1.Transformation)(nil),
                    Traffic:        v1.SubscriberTraffic{},
                    Security:       (*v1.SubscriberSecurity)(nil),
                },
                Event: (*v1.EventSubscription)(nil),
            },
        },
    },
    Status: v1.RoverStatus{},
}
---

[Rover Mapper MapExposures must map exposures correctly - 1]
&v1.Rover{
    TypeMeta:   v1.TypeMeta{},
    ObjectMeta: v1.ObjectMeta{},
    Spec:       v1.RoverSpec{
        Zone:           "",
        IpRestrictions: (*v1.IpRestrictions)(nil),
        ClientSecret:   "",
        Exposures:      {
            {
                Api: &v1.ApiExposure{
                    BasePath:  "/eni/distr/v1",
                    Upstreams: {
                        {URL:"https://httpbin.org/anything", Weight:0},
                    },
                    Visibility: "World",
                    Approval:   v1.Approval{
                        Strategy:     "Simple",
                        TrustedTeams: nil,
                    },
                    Transformation: (*v1.Transformation)(nil),
                    Traffic:        v1.Traffic{},
                    Security:       (*v1.Security)(nil),
                },
                Event: (*v1.EventExposure)(nil),
            },
        },
        Subscriptions: nil,
    },
    Status: v1.RoverStatus{},
}
---

[Rover Mapper MapSubscriptions must map subscriptions correctly - 1]
&v1.Rover{
    TypeMeta:   v1.TypeMeta{},
    ObjectMeta: v1.ObjectMeta{},
    Spec:       v1.RoverSpec{
        Zone:           "",
        IpRestrictions: (*v1.IpRestrictions)(nil),
        ClientSecret:   "",
        Exposures:      nil,
        Subscriptions:  {
            {
                Api: &v1.ApiSubscription{
                    BasePath:       "/eni/distr/v1",
                    Organization:   "",
                    Transformation: (*v1.Transformation)(nil),
                    Traffic:        v1.SubscriberTraffic{},
                    Security:       (*v1.SubscriberSecurity)(nil),
                },
                Event: (*v1.EventSubscription)(nil),
            },
        },
    },
    Status: v1.RoverStatus{},
}
---

[Rover Mapper MapRequest must map a RoverUpdateRequest to a Rover correctly - 1]
&v1.Rover{
    TypeMeta:   v1.TypeMeta{Kind:"Rover", APIVersion:"rover.cp.ei.telekom.de/v1"},
    ObjectMeta: v1.ObjectMeta{
        Name:                       "rover-local-sub",
        GenerateName:               "",
        Namespace:                  "poc--poc--eni--hyperion",
        SelfLink:                   "",
        UID:                        "",
        ResourceVersion:            "",
        Generation:                 0,
        CreationTimestamp:          time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC),
        DeletionTimestamp:          (*v1.Time)(nil),
        DeletionGracePeriodSeconds: (*int64)(nil),
        Labels:                     {"cp.ei.telekom.de/environment":"poc"},
        Annotations:                {},
        OwnerReferences:            nil,
        Finalizers:                 nil,
        ManagedFields:              nil,
    },
    Spec: v1.RoverSpec{
        Zone:           "zone",
        IpRestrictions: (*v1.IpRestrictions)(nil),
        ClientSecret:   "",
        Exposures:      {
            {
                Api: &v1.ApiExposure{
                    BasePath:  "/eni/distr/v1",
                    Upstreams: {
                        {URL:"https://httpbin.org/anything", Weight:0},
                    },
                    Visibility: "World",
                    Approval:   v1.Approval{
                        Strategy:     "Simple",
                        TrustedTeams: nil,
                    },
                    Transformation: (*v1.Transformation)(nil),
                    Traffic:        v1.Traffic{},
                    Security:       (*v1.Security)(nil),
                },
                Event: (*v1.EventExposure)(nil),
            },
        },
        Subscriptions: {
            {
                Api: &v1.ApiSubscription{
                    BasePath:       "/eni/distr/v1",
                    Organization:   "",
                    Transformation: (*v1.Transformation)(nil),
                    Traffic:        v1.SubscriberTraffic{},
                    Security:       (*v1.SubscriberSecurity)(nil),
                },
                Event: (*v1.EventSubscription)(nil),
            },
        },
    },
    Status: v1.RoverStatus{},
}
---

[Rover Mapper MapRequest must return an error if the input rover is nil - 1]
&v1.Rover{
    TypeMeta:   v1.TypeMeta{Kind:"Rover", APIVersion:"rover.cp.ei.telekom.de/v1"},
    ObjectMeta: v1.ObjectMeta{
        Name:                       "rover-local-sub",
        GenerateName:               "",
        Namespace:                  "poc--poc--eni--hyperion",
        SelfLink:                   "",
        UID:                        "",
        ResourceVersion:            "",
        Generation:                 0,
        CreationTimestamp:          time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC),
        DeletionTimestamp:          (*v1.Time)(nil),
        DeletionGracePeriodSeconds: (*int64)(nil),
        Labels:                     {"cp.ei.telekom.de/environment":"poc"},
        Annotations:                {},
        OwnerReferences:            nil,
        Finalizers:                 nil,
        ManagedFields:              nil,
    },
    Spec:   v1.RoverSpec{},
    Status: v1.RoverStatus{},
}
---
